File: src/main/java/leet_code/Dota2Senate.java
package leet_code;

import java.util.ArrayDeque;
import java.util.Deque;

public class Dota2Senate {

    public static void main(String[] args) {

        System.out.println(predictPartyVictory("RD"));

    }


    public static String predictPartyVictory(String senate) {

        Deque<Integer> radiant = new ArrayDeque<>();

        Deque<Integer> dire = new ArrayDeque<>();

        for (int i = 0; i < senate.length(); i++) {
            if (senate.charAt(i) == 'R')
                radiant.add(i);
            else dire.add(i);
        }

        int counter = senate.length();

        while (!radiant.isEmpty() && !dire.isEmpty()) {
            int rad = radiant.peekFirst();
            int dir = dire.peekFirst();
            counter++;
            if (rad > dir) { // Dire win the battle. Dire goes to next round
                dire.addLast(counter);
                dire.pop();
                radiant.pop();
            }else {// Radiant win the battle. Radiant goes to next round
                radiant.addLast(counter);
                radiant.pop();
                dire.pop();
            }
        }

        return radiant.isEmpty() ? "Dire" : "Radiant";
    }


}

File: src/main/java/leet_code/may2024/Dota2Senate.go
package main

import "fmt"

func main() {
	fmt.Println(predictPartyVictory("RDD"))
	fmt.Println()
	fmt.Println(predictPartyVictory("RD"))
}

func predictPartyVictory(senate string) string {

	var rad []int
	var dir []int

	for i, s := range senate {

		if s == 'R' {
			rad = append(rad, i)
		} else {
			dir = append(dir, i)
		}
	}

	index := len(senate)

	for len(dir) > 0 && len(rad) > 0 {
		index++
		d := dir[0]
		dir = dir[1:]

		r := rad[0]
		rad = rad[1:]

		if r < d {
			rad = append(rad, index)
		} else {
			dir = append(dir, index)
		}
	}

	if len(dir) > 0 {
		return "Dire"
	} else {
		return "Radian"
	}
}
